evm/src/fixed_recursive_verifier.rs

```diff
@@ -0,0 +1,398 @@
+use std::collections::BTreeMap;
+use std::ops::Range;
+
+use itertools::Itertools;
+use plonky2::field::extension::Extendable;
+use plonky2::hash::hash_types::RichField;
+use plonky2::hash::hashing::SPONGE_WIDTH;
+use plonky2::iop::challenger::RecursiveChallenger;
+use plonky2::iop::target::Target;
+use plonky2::iop::witness::{PartialWitness, WitnessWrite};
+use plonky2::plonk::circuit_builder::CircuitBuilder;
+use plonky2::plonk::circuit_data::{CircuitConfig, CircuitData};
+use plonky2::plonk::config::{AlgebraicHasher, GenericConfig, Hasher};
+use plonky2::plonk::proof::{ProofWithPublicInputs, ProofWithPublicInputsTarget};
+use plonky2::util::timing::TimingTree;
+
+use crate::all_stark::{all_cross_table_lookups, AllStark, Table, NUM_TABLES};
+use crate::config::StarkConfig;
+use crate::cpu::cpu_stark::CpuStark;
+use crate::cross_table_lookup::{verify_cross_table_lookups_circuit, CrossTableLookup};
+use crate::generation::GenerationInputs;
+use crate::keccak::keccak_stark::KeccakStark;
+use crate::keccak_sponge::keccak_sponge_stark::KeccakSpongeStark;
+use crate::logic::LogicStark;
+use crate::memory::memory_stark::MemoryStark;
+use crate::permutation::{get_grand_product_challenge_set_target, GrandProductChallengeSet};
+use crate::proof::StarkProofWithMetadata;
+use crate::prover::prove;
+use crate::recursive_verifier::{
+    add_common_recursion_gates, recursive_stark_circuit, PlonkWrapperCircuit, PublicInputs,
+    StarkWrapperCircuit,
+};
+use crate::stark::Stark;
+
+/// The recursion threshold. We end a chain of recursive proofs once we reach this size.
+const THRESHOLD_DEGREE_BITS: usize = 13;
```

Comment:
I was having trouble shrinking to 2^12 since these shrinking circuits have a bunch of constants for the verification key, taking a bit of extra witness space. We might be able to make it work, but I think it's not that important, since 2^13 is still small compared to the initial wrapper proofs (which check ~80 queries) or the STARKs themselves.